include/aspect/assembly.h

```diff
@@ -365,6 +375,42 @@ namespace aspect
       struct AssemblerLists
       {
         /**
+         * A structure used to accumulate the results of the
+         * compute_advection_system_residual slot functions below.
+         * It takes an iterator range of vectors and returns the element-wise
+         * sum of the values.
+         */
+        template<typename T>
+        struct ResidualWeightSum
+        {
+          typedef T result_type;
+
+          template<typename InputIterator>
+          T operator()(InputIterator first, InputIterator last) const
+          {
+            // If there are no slots to call, just return the
+            // default-constructed value
+            if (first == last)
+              return T();
+
+            T sum = *first++;
+            while (first != last)
+              {
+                Assert(sum.size() == first->size(),
+                       ExcMessage("Invalid vector length for residual summation."));
+
+                for (unsigned int i = 0; i < first->size(); ++i)
+                  sum[i] += (*first)[i];
```

Comment:
You're using the assumption here that T is in fact a vector type. If so, I think the class may be clearer if you call T `VectorType`.